File: solutions/elixir/protein-translation/1/protein_translation.exs
defmodule ProteinTranslation do
  @codons %{
    :UGU => "Cysteine",
    :UGC => "Cysteine",
    :UUA => "Leucine",
    :UUG => "Leucine",
    :AUG => "Methionine",
    :UUU => "Phenylalanine",
    :UUC => "Phenylalanine",
    :UCU => "Serine",
    :UCC => "Serine",
    :UCA => "Serine",
    :UCG => "Serine",
    :UGG => "Tryptophan",
    :UAU => "Tyrosine",
    :UAC => "Tyrosine",
    :UAA => "STOP",
    :UAG => "STOP",
    :UGA => "STOP"
  }

  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {atom, list(String.t())}
  def of_rna(rna) do
    protein_list = for <<codon::binary-3 <- rna>> do
      {:ok, protein} = of_codon codon
      protein
    end
    {:ok, protein_list}
  end

  defp chunk_rna(codon) do
    for <<c::binary-3 <- codon>> do
      {:ok, protein} = of_codon(c)
      protein
    end
  end

  @doc """
  Given a codon, return the corresponding protein

  UGU -> Cysteine
  UGC -> Cysteine
  UUA -> Leucine
  UUG -> Leucine
  AUG -> Methionine
  UUU -> Phenylalanine
  UUC -> Phenylalanine
  UCU -> Serine
  UCC -> Serine
  UCA -> Serine
  UCG -> Serine
  UGG -> Tryptophan
  UAU -> Tyrosine
  UAC -> Tyrosine
  UAA -> STOP
  UAG -> STOP
  UGA -> STOP
  """
  @spec of_codon(String.t()) :: {atom, String.t()}
  def of_codon(codon), do: Map.fetch(@codons, String.to_atom(codon))
end

File: solutions/csharp/protein-translation/1/ProteinTranslation.cs
enum Protein
{
    Methionine,
    Phenylalanine,
    Leucine,
    Serine,
    Tyrosine,
    Cysteine,
    Tryptophan,
    Stop,
}

static class ProteinExtensions
{
    public static Protein FromCodon(string codon) => codon switch
    {
        "AUG" => Protein.Methionine,
        "UUU" or "UUC" => Protein.Phenylalanine,
        "UUA" or "UUG" => Protein.Leucine,
        "UCU" or "UCC" or "UCA" or "UCG" => Protein.Serine,
        "UAU" or "UAC" => Protein.Tyrosine,
        "UGU" or "UGC" => Protein.Cysteine,
        "UGG" => Protein.Tryptophan,
        "UAA" or "UAG" or "UGA" => Protein.Stop,
        _ => throw new ArgumentException($"Unrecognized codon: {codon}", nameof(codon))
    };
}

public static class ProteinTranslation
{
    public static string[] Proteins(string strand)
    {
        if (strand.Length % 3 != 0)
        {
            throw new ArgumentException("strand not divisible by 3");
        }

        var codons = strand.Chunk(3).Select(c => new string(c));

        var proteins = new List<string>();

        foreach (var codon in codons)
        {
            var protein = ProteinExtensions.FromCodon(codon);
            if (protein == Protein.Stop) break;
            
            proteins.Add(protein.ToString());
        }
        
        return proteins.ToArray();
    }
    
}